app/src/androidTest/java/com/github/dedis/student20_pop/utility/network/ProtocolTest.java

```diff
@@ -0,0 +1,205 @@
+package com.github.dedis.student20_pop.utility.network;
+
+import com.github.dedis.student20_pop.model.Person;
+
+import net.jodah.concurrentunit.Waiter;
+
+import org.glassfish.tyrus.server.Server;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.Collections;
+import java.util.Locale;
+import java.util.concurrent.TimeoutException;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import javax.websocket.DeploymentException;
+import javax.websocket.OnMessage;
+import javax.websocket.Session;
+import javax.websocket.server.ServerEndpoint;
+
+/**
+ * Test that the protocol handles correctly and sends messages.
+ * Also test the CompletableFuture system and timeouts
+ */
+public class ProtocolTest {
+
+    private Server startAcceptEverythingServer() throws DeploymentException {
+        Server server = new Server("localhost", 2020, "", PerfectServer.class);
+        server.start();
+        return server;
+    }
+
+    // Test timeouts
+    private Server startNoAnswerServer() throws DeploymentException {
+        Server server = new Server("localhost", 2020, "", TimeoutServer.class);
+        server.start();
+        return server;
+    }
+
+    @Test
+    public void testTimeout() throws DeploymentException, TimeoutException, InterruptedException {
+        Server server = startNoAnswerServer();
+        Waiter waiter = new Waiter();
+
+        Person bob = new Person("Bob");
+        HighLevelClientProxy proxy = PoPClientEndpoint.connectToServer(URI.create("ws://localhost:2020/"), bob);
+
+        proxy.createLoa("name", 0, 0, bob.getId())
+            .whenComplete((i, t) -> {
+                waiter.assertTrue(t != null);
+                waiter.resume();
+            });
+
+        synchronized (this) {
+            wait(6000);
+        }
+
+        proxy.lowLevel().purge();
+
+        waiter.await(10000, 1);
```

Comment:
You should put 10000 in a variable as you use it in almost every test

Reply:
Done !